dcc-portal-ui/app/scripts/pathwayviewer/js/services.js

```diff
@@ -1213,4 +1213,254 @@ angular.module('icgc.pathwayviewer.directives.services', [])
+      return GeneSetHierarchy.uiPathwayHierarchy(geneSet.hierarchy, geneSet);
+    };
+    _pathwayDataService.getParentPathwayId = function (geneSet) {
+      return _pathwayDataService.getUIParentPathways(geneSet)[0].geneSetId;
+    };
+    _pathwayDataService.getPathwayId = function (geneSet) {
+      return _pathwayDataService.getUIParentPathways(geneSet)[0].diagramId;
+    };
+    _pathwayDataService.getPathwayXML = function (geneSet) {
+      return GeneSetService.getPathwayXML(_pathwayDataService.getPathwayId(geneSet)).then(function (pathwayXML) {
+        return pathwayXML;
+      });
+    };
+    _pathwayDataService.getPathwayZoom = function (geneSet) {
+      if (_pathwayDataService.getParentPathwayId(geneSet) === _pathwayDataService.getPathwayId(geneSet)) {
+        return $q.resolve;
+      }
+ 
+     return GeneSetService.getPathwayZoom(_pathwayDataService.getParentPathwayId(geneSet))
+       .then(function (pathwayZoom) {
+         return pathwayZoom;
+       });
+    };
+    _pathwayDataService.getPathwayProteinMap = function (geneSet) {
+    return GeneSetService.getPathwayProteinMap(_pathwayDataService.getParentPathwayId(geneSet), [])
+      .then(function (pathwayProteinMap) {
+        return pathwayProteinMap;
+      });
+    };
+    _pathwayDataService.getEntitySetId = function (geneSet) {
+      return SetService.getTransientSet('GENE', {
+        'filters': {gene:{pathwayId:{is:[_pathwayDataService.getParentPathwayId(geneSet)]}}},
+        'sortBy': 'id',
+        'sortOrder': 'ASCENDING',
+        'name': _pathwayDataService.getParentPathwayId(geneSet),
+        'size': geneSet.geneCount,
+        'transient': true 
+      }).then(function (entitySetData) {
+        return entitySetData.id;
+      });
+    };
+    _pathwayDataService.getDrugs = function (entitySetId) {
+      return CompoundsService.getCompoundsFromEntitySetId(entitySetId).then(function(drugs) {  
+        return drugs;
+      });
+    };
+    _pathwayDataService.getMutationHighlights = function (pathwayProteinMap) {
+      var pathwayMutationHighlights = [];
+      _.forEach(pathwayProteinMap, function (value, id) {
+        if (value && value.dbIds) {
+          var dbIds = value.dbIds.split(',');
+          pathwayMutationHighlights.push({
+            uniprotId: id,
+            dbIds: dbIds,
+            value: value.value
+          });
+        }
+      });
+      return pathwayMutationHighlights;
+    };
+    _pathwayDataService.getDrugHighlights = function (drugs, pathwayProteinMap) {
+      var drugMap = {};
+      _.forEach(drugs, function(drug) {
+        _.forEach(drug.genes, function(gene) {
+          var uniprotId = gene.uniprot;
+
+          if (_.isUndefined(drugMap[uniprotId])) {
+              drugMap[uniprotId] = [];
+          }
+
+          drugMap[uniprotId].push({
+            'zincId': drug.zincId,
+            'name': drug.name
+          });
+        });
+      });
```

Comment:
Could you try replacing this

```js
var drugMap = {};
      _.forEach(drugs, function(drug) {
        _.forEach(drug.genes, function(gene) {
          var uniprotId = gene.uniprot;

          if (_.isUndefined(drugMap[uniprotId])) {
              drugMap[uniprotId] = [];
          }

          drugMap[uniprotId].push({
            'zincId': drug.zincId,
            'name': drug.name
          });
        });
      });
```
with this
```js

var drugMap = _(drugs)
  .map('genes')
  .flatten()
  .groupBy('uniprotId')
  .value()
```

And let me know if it still behaves as expected?